src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,791 @@
+  where
+  lt = even<odd x<y
+
+... | tri≈ _   x≡y _ =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  1+2x≡1+2y = cong 1+[2_] x≡y;  1+2x<2[1+x] = 1+[2x]<2[1+x] x
+
+  gt : 1+[2 y ] < 2[1+ x ]
+  gt = subst (_< 2[1+ x ]) 1+2x≡1+2y (1+[2x]<2[1+x] x)
+
+... | tri> _ _ y<x =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt : 1+[2 y ] < 2[1+ x ]
+  gt = odd<even (inj₁ y<x)
+
+<-cmp 1+[2 _ ] zero =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt = 0<odd
+
+<-cmp 1+[2 x ] 2[1+ y ]  with <-cmp x y
+... | tri< x<y _ _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)
+  where
+  lt = odd<even (inj₁ x<y)
+
+... | tri≈ _ x≡y _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)
+  where
+  lt = odd<even (inj₂ x≡y)
+
+... | tri> _ _ x>y =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt = even<odd x>y
+
+<-cmp 1+[2 x ] 1+[2 y ]  with <-cmp x y
+... | tri< x<y _ _ =  tri< lt (<⇒≢ lt) (<⇒≯ lt)
+  where
+  lt = odd<odd x<y
+
+... | tri≈ _ x≡y _ =  tri≈ (≡⇒≮ eq) eq (≡⇒≯ eq)
+  where
+  eq = cong 1+[2_] x≡y
+
+... | tri> _ _ x>y =  tri> (>⇒≮ gt) (>⇒≢ gt) gt
+  where
+  gt = odd<odd x>y
+
+------------------------------------------------------------------------------
+-- Several monotonicity proofs for toℕ, fromℕ.
+------------------------------------------------------------------------------
+
+toℕ-mono-< :  toℕ Preserves _<_ ⟶ ℕ._<_
+toℕ-mono-< {_}        {zero}     ()
+toℕ-mono-< {zero}     {2[1+ _ ]} _               =  ℕp.0<1+n
+toℕ-mono-< {zero}     {1+[2 _ ]} _               =  ℕp.0<1+n
+toℕ-mono-< {2[1+ x ]} {2[1+ y ]} (even<even x<y) =  x'N<y'N
+  where
+  open ℕp.≤-Reasoning
+  xN      = toℕ x
+  yN      = toℕ y
+  xN<yN   = toℕ-mono-< x<y
+  x'N<y'N = begin
+    ℕ.suc (2 ℕ.* (ℕ.suc xN))   ≤⟨ ℕp.+-monoʳ-≤ 1 (ℕp.*-monoʳ-≤ 2 xN<yN) ⟩
+    ℕ.suc (2 ℕ.* yN)            ≤⟨ ℕp.≤-step ℕp.≤-refl ⟩
+    2 ℕ.+ (2 ℕ.* yN)            ≡⟨ sym (ℕp.*-distribˡ-+ 2 1 yN) ⟩
+    2 ℕ.* (ℕ.suc yN)            ∎
+
+toℕ-mono-< {2[1+ x ]} {1+[2 y ]} (even<odd x<y) =  x'N<y'N
+  where
+  xN      = toℕ x
+  yN      = toℕ y
+  xN<yN   = toℕ-mono-< x<y
+  x'N<y'N = ℕp.+-monoʳ-≤ 1 (ℕp.*-monoʳ-≤ 2 xN<yN)
+
+toℕ-mono-< {1+[2 x ]} {2[1+ y ]} (odd<even (inj₁ x<y)) =  x'N<y'N
+  where
+  open ℕp.≤-Reasoning
+  xN      = toℕ x
+  yN      = toℕ y
+  xN<yN   = toℕ-mono-< x<y
+  x'N<y'N = begin
+    ℕ.suc (ℕ.suc (2 ℕ.* xN))   ≡⟨ refl ⟩
+    2 ℕ.+ (2 ℕ.* xN)            ≡⟨ sym (ℕp.*-distribˡ-+ 2 1 xN) ⟩
+    2 ℕ.* (ℕ.suc xN)            ≤⟨ ℕp.*-monoʳ-≤ 2 xN<yN ⟩
+    2 ℕ.* yN                     ≤⟨ ℕp.*-monoʳ-≤ 2 (ℕp.≤-step ℕp.≤-refl) ⟩
+    2 ℕ.* (ℕ.suc yN)            ∎
+
+
+toℕ-mono-< {1+[2 x ]} {2[1+ .x ]} (odd<even (inj₂ refl)) =  x'N<y'N
+  where
+  xN = toℕ x
+
+  x'N<y'N :  ℕ.suc (ℕ.suc (2 ℕ.* xN))  ℕ.≤  2 ℕ.* (ℕ.suc xN)
+  x'N<y'N =  ℕp.≤-reflexive (sym (ℕp.*-distribˡ-+ 2 1 xN))
+
+toℕ-mono-< {1+[2 x ]} {1+[2 y ]} (odd<odd x<y) =  x'N<y'N
+  where
+  xN      = toℕ x
+  yN      = toℕ y
+  xN<yN   = toℕ-mono-< x<y
+  x'N<y'N = ℕp.+-monoʳ-< 1 (ℕp.*-monoʳ-< 1 xN<yN)
+
+
+fromℕ-mono-< :  fromℕ Preserves ℕ._<_ ⟶ _<_
+fromℕ-mono-< {m} {n} m<n =
+  let
```

Comment:
Please could you use `where` instead of `let`

Reply:
```
    x       = fromℕ m;      y       = fromℕ n
    toℕ-x≡m = toℕ-fromℕ m;  toℕ-y≡n = toℕ-fromℕ n
  in
  case <-cmp x y of \
```
> Is there a reason why you're using case instead of with?

I have rewritten it to
```
fromℕ-mono-< {m} {n} m<n =  aux (<-cmp x y)
  where
  open P.≡-Reasoning
  x       = fromℕ m;      y       = fromℕ n
  toℕ-x≡m = toℕ-fromℕ m;  toℕ-y≡n = toℕ-fromℕ n

  aux : Tri (x < y) (x ≡ y) (x > y) → x < y
  aux (tri< x<y _   _) =  x<y
  ...
```
But I do not know how write this via ``with`` so that the four assignments
after ``where`` to be common for all branches of ``with``.
Can you demonstrate?

Reply:
Well I would usually use another `with`.
```
fromℕ-mono-< {m} {n} m<n with toℕ-fromℕ m | toℕ-fromℕ n
... | toℕ[x]≡m | toℕ[y]≡n = ...
```
but I recognise that may not be very efficient...

Reply:
As you are in a doubt, I do not change this place.